File: app/src/main/java/com/example/translator_three/AppDatabase.java
//package com.example.translator_three;
//
//import android.content.Context;
//import androidx.room.Database;
//import androidx.room.Room;
//import androidx.room.RoomDatabase;
//
//import com.example.translator_three.dao.TranslationCacheDao;
//import com.example.translator_three.dao.UserDao;
//import com.example.translator_three.model.TranslationCache;
//import com.example.translator_three.model.User;
//
//@Database(
//        entities = {User.class, TranslationCache.class},
//        version = 2,  // 已更新为2
//        exportSchema = false
//)
//public abstract class AppDatabase extends RoomDatabase {
//    //  volatile 关键字确保多线程可见性（之前缺失，补充上）
//    private static  volatile AppDatabase INSTANCE;
//
//    public abstract UserDao userDao();
//    public abstract TranslationCacheDao translationCacheDao();
//
//    public static synchronized AppDatabase getDatabase(Context context) {
//        if (INSTANCE == null) {
//            INSTANCE = Room.databaseBuilder(
//                            context.getApplicationContext(),
//                            AppDatabase.class,
//                            "translator_db"
//                    )
//                    .allowMainThreadQueries()
//                    // 关键：添加允许破坏性迁移
//                    .fallbackToDestructiveMigration()
//                    .build();
//        }
//        return INSTANCE;
//    }
//
//    // 提供销毁实例方法（方便测试时重置）
//    public static void destroyInstance() {
//        INSTANCE = null;
//    }
//}



package com.example.translator_three;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.example.translator_three.dao.UserDao; // 只保留 UserDao
import com.example.translator_three.model.User; // 只保留 User 实体

// 核心修改1：entities 仅保留 User.class，移除 TranslationCache.class
// 核心修改2：版本号可保留 2（也可重置为 1，建议重置避免旧版本冲突）
@Database(
        entities = {User.class},
        version = 1,  // 重置为 1，彻底清空旧的冲突数据
        exportSchema = false
)
public abstract class AppDatabase extends RoomDatabase {
    private static volatile AppDatabase INSTANCE;

    // 核心修改3：只保留 UserDao，移除 TranslationCacheDao
    public abstract UserDao userDao();

    public static synchronized AppDatabase getDatabase(Context context) {
        if (INSTANCE == null) {
            INSTANCE = Room.databaseBuilder(
                            context.getApplicationContext(),
                            AppDatabase.class,
                            "user_db" // 已改为 user_db，专属用户数据
                    )
                    .allowMainThreadQueries()
                    .fallbackToDestructiveMigration()
                    .build();
        }
        return INSTANCE;
    }

    public static void destroyInstance() {
        INSTANCE = null;
    }
}

File: app/src/main/java/com/example/translator_three/database/AppDatabase.kt
package com.example.translator_three.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.example.translator_three.dao.TranslationCacheDao
import com.example.translator_three.model.TranslationCache

//Room组件之一，数据库类，用于保存数据库并作为应用持久性数据底层连接的主要访问点。
// 数据库版本（升级时修改）
@Database(entities = [TranslationCache::class], version = 1, exportSchema = false)
abstract class AppDatabase : RoomDatabase() {
    // 提供 DAO 实例
    abstract fun translationCacheDao(): TranslationCacheDao

    // Kotlin 单例（双重校验锁），避免重复建立数据库连接
    //Java的static 变量/方法	= companion object { } 内的成员

//    companion object {
//        @Volatile
//        private var INSTANCE: AppDatabase? = null
//
//        fun getInstance(context: Context): AppDatabase {
//            return INSTANCE ?: synchronized(this) {
//                val instance = Room.databaseBuilder(
//                    context.applicationContext,
//                    AppDatabase::class.java,
//                    "translator_db"
//                )
//                    // 开发阶段允许清空数据库（上线前删除）
//                    // .fallbackToDestructiveMigration()
//                    // 核心2：强制破坏性迁移（清空所有旧数据，重建表）
//                    .fallbackToDestructiveMigration()
//                    // 额外：兼容所有版本的破坏性迁移（兜底）
//                    .fallbackToDestructiveMigrationOnDowngrade()
//
//                    .build()
//                INSTANCE = instance
//                instance
//            }
//        }
//    }

    //这个才是双重检查好不，有时间测试了的话要是运行没错就把上面那注释删了吧，我见代码后面有变化
    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase {
            // 第一次检查（无锁）
            return INSTANCE ?: synchronized(this) {
                // 第二次检查（锁内，防止其他线程已初始化）
                INSTANCE ?: run {
                    Room.databaseBuilder(
                        context.applicationContext,
                        AppDatabase::class.java,
                        "translator_db"
                    )
                        .fallbackToDestructiveMigration()  // 开发阶段允许清空数据库（上线前删除）（清空所有旧数据，重建表）
                        .fallbackToDestructiveMigrationOnDowngrade()  // 额外：兼容所有版本的破坏性迁移（兜底）
                        .build()
                        .also { INSTANCE = it }  // 赋值并返回
                }
            }
        }
    }
}